src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ResourceDataNode.cs

```diff
@@ -0,0 +1,182 @@
+        }
+
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        public ISymbolNode EndSymbol => _endSymbol;
+
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => ObjectNodeSection.ReadOnlyDataSection;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        public int Offset => 0;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__embedded_resourcedata");
+        }
+
+        protected override string GetName() => this.GetMangledName();
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node has no relocations.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+            
+            byte[] blob = GenerateResourceBlob(factory.CompilationModuleGroup, factory.TypeSystemContext);
+            return new ObjectData(
+                blob,
+                Array.Empty<Relocation>(),
+                1,
+                new ISymbolNode[]
+                {
+                                this,
+                                EndSymbol
+                });
+        }
+
+        /// <summary>
+        /// Extracts resources from all modules being compiled into a single blob and saves
+        /// the information needed to create an index into that blob.
+        /// </summary>
+        private byte[] GenerateResourceBlob(CompilationModuleGroup compilationModuleGroup, CompilerTypeSystemContext typeSystemContext)
+        {
+            int totalLength = 0;
+
+            // Build up index information
+            foreach (EcmaModule module in _modulesSeen.OfType<EcmaModule>())
+            {
+                PEMemoryBlock resourceDirectory = module.PEReader.GetSectionData(module.PEReader.PEHeaders.CorHeader.ResourcesDirectory.RelativeVirtualAddress);
+
+                foreach (var resourceHandle in module.MetadataReader.ManifestResources)
+                {
+                    ManifestResource resource = module.MetadataReader.GetManifestResource(resourceHandle);
+                    string resourceName = module.MetadataReader.GetString(resource.Name);
+                    string assemblyName = GetAssemblyName(module);
+                    BlobReader reader = resourceDirectory.GetReader(checked((int)resource.Offset), resourceDirectory.Length - (int)resource.Offset);
+                    int length = checked((int)reader.ReadUInt32());
+                    ResourceIndexData indexData = new ResourceIndexData(assemblyName, resourceName, totalLength, (int)resource.Offset + sizeof(Int32), module, length);
+                    IndexData.Add(indexData);
+                    totalLength += length;
```

Comment:
Made the entire block checked for better readability and in case I missed anything else.